aesara/tensor/basic.py

```diff
@@ -1812,19 +1814,163 @@ def extract_constant(x, elemwise=True, only_process_constants=False):
-        axes = list(range((x.ndim - 1), -1, -1))
-    ret = DimShuffle(x.broadcastable, axes)(x)
-    if x.name and axes == list(range((x.ndim - 1), -1, -1)):
-        ret.name = x.name + ".T"
-    return ret
+    def make_node(self, input_arr, new_order=None):
+        input_arr = as_tensor_variable(input_arr)
+        ib = list(input_arr.type.broadcastable)
+
+        if new_order is None:
+            new_order = as_tensor_variable(np.flip(np.arange(input_arr.ndim)))
+            ib.reverse()
+            output = TensorType(dtype=input_arr.type.dtype, shape=tuple(ib))()
+            if input_arr.name:
+                output.name = input_arr.name + ".T"
+        else:
+            from aesara.tensor.var import TensorConstant
+
+            assert isinstance(new_order, TensorConstant)
+            # TODO: Try to remove this eval
+            _new_order = list(new_order.eval())
+            ob = []
+            if len(_new_order) != len(ib):
+                raise ValueError(
+                    f"Expected {len(ib)} number of values, found {len(_new_order)}."
+                )
+            for a in _new_order:
+                if a >= len(ib):
+                    raise ValueError("Found value greater than number of axes.")
+                if a < 0:
+                    raise ValueError(
+                        "Negative reference to axis, (e.g. axis = -1) is not allowed."
+                    )
+                ob.append(ib[a])
+            output = TensorType(dtype=input_arr.type.dtype, shape=ob)()
+
+        return Apply(self, [input_arr, new_order], [output])
+
+    def perform(self, node, inputs_storage, output_storage):
+        (input_arr, new_order) = inputs_storage
+        (storage,) = output_storage
+        res = input_arr.transpose(new_order)
+        storage[0] = np.asarray(res)
+
+    def grad(self, inp, grads):
+        from aesara.tensor.basic import as_tensor_variable
+
+        (x, axis) = inp
+        (gz,) = grads
+        gz = as_tensor_variable(gz)
+        grad_order = [-1] * len(x.type.broadcastable)
+        for i, v in enumerate(axis.data):
+            grad_order[v] = i
+
+        return [
+            TransposeOp()(
+                Elemwise(scalar_identity)(gz), as_tensor_variable(grad_order)
+            ),
+            as_tensor_variable(grad_order).astype(dtype=config.floatX),
+        ]
+
+    def R_op(self, inputs, eval_points):
+        # TransposeOp is not differentiable wrt to its axes, therefore we
+        # do not even need to consider the eval_points provided for those
+        # (they should be defaulted to zeros_like by the global R_op)
+        if eval_points[0] is None:
+            return [None]
+        return self(eval_points[0], *inputs[1:], return_list=True)
+
+    def infer_shape(self, fgraph, node, shapes):
+        ishp = shapes[0]
+        rval = [ishp[i] for i in node.inputs[1].data]
+        return [rval]
+
+
+transpose = TransposeOp()
+
+
+class SqueezeDimsOp(Op):
```

Comment:
This is a regression if we don't have a `grad` and `infer_shape` methods... Also there are probably rewrites that need to be updated to consider squeezing Ops as well?